Fix register constant to be usable in exportz

tugsbayasgalan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #147533



cc @ezyang @SherlockNoMad @EikanWang @jgong5 @wenzhe-nrv

Differential Revision: [D69939737](https://our.internmc.facebook.com/intern/diff/D69939737)

@diff-train-skip-merge

tugsbayasgalan committed:
Fix register constant to be usable in exportz

[ghstack-poisoned]

zou3519 commented on torch/fx/_pytree.py:
```diff
@@ -36,30 +36,23 @@ def _deregister_pytree_flatten_spec(
+            flat = tree_flatten_spec(child, child_spec)
+            result += flat
+        return result
+    return list(tree_flatten(pytree)[0])
```
Can you add a check that the spec after flattening is the same as the spec being passed in?

If it is the same, then good. If it is different, we should raise an error.

tugsbayasgalan replied:
Updated!

tugsbayasgalan committed:
Update on "Fix register constant to be usable in exportz"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69939737).

zou3519 commented on torch/fx/_pytree.py:
```diff
@@ -36,30 +36,29 @@ def _deregister_pytree_flatten_spec(
-            f"{type(pytree)} does not have a flatten_fn_spec associated with it. Please register one with "
-            "torch.fx._pytree.register_pytree_flatten_spec.  If you have serialized your model, make "
-            "sure that any custom pytrees have been registered before loading it.",
+            "Real spec {real_spec} of object {pytree} is different from expected spec {spec}. "
```
Needs f-string

zou3519 commented on torch/fx/_pytree.py:
```diff
@@ -36,30 +36,29 @@ def _deregister_pytree_flatten_spec(
-        flat = tree_flatten_spec(child, child_spec, exact_structural_match)
-        result += flat
-    return result
+    return list(flat_result) if not isinstance(flat_result, list) else flat_result
```
flat_result should always be a list

tugsbayasgalan committed:
Update on "Fix register constant to be usable in exportz"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69939737](https://our.internmc.facebook.com/intern/diff/D69939737)

[ghstack-poisoned]

tugsbayasgalan commented:
@pytorchbot rebase

tugsbayasgalan committed:
Update

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix register constant to be usable in exportz"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69939737](https://our.internmc.facebook.com/intern/diff/D69939737)

[ghstack-poisoned]

tugsbayasgalan committed:
Update on "Fix register constant to be usable in exportz"




cc ezyang SherlockNoMad EikanWang jgong5 wenzhe-nrv

Differential Revision: [D69939737](https://our.internmc.facebook.com/intern/diff/D69939737)

[ghstack-poisoned]

tugsbayasgalan commented:
@tugsbayasgalan has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D69939737).
